Tiller Pod is Crashed when values yaml have issue.

Logs are bellow
```
[tiller] 2017/12/21 22:25:30 rendering consume-sample chart using values
panic: reflect: Zero(nil) [recovered]
	panic: reflect: Zero(nil)
 goroutine 313 [running]:
text/template.errRecover(0xc420d53098)
	/usr/local/go/src/text/template/exec.go:146 +0x62
panic(0x1b48260, 0xc4203bfa00)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
reflect.Zero(0x0, 0x0, 0x0, 0x1ba7680, 0xc4203bf9f0)
	/usr/local/go/src/reflect/value.go:2122 +0x134
text/template.(*state).validateType(0xc420d53018, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc420d52c28)
	/usr/local/go/src/text/template/exec.go:697 +0x700
text/template.(*state).evalArg(0xc420d53018, 0x1cc8440, 0xc420231020, 0x15, 0x0, 0x0, 0x2ab1ae0, 0xc4208aa000, 0x6, 0x4810a6, ...)
	/usr/local/go/src/text/template/exec.go:746 +0x870
```

Proposal, Better to have helm lint before installing chart. So Error can be handled gracefully instead of Tiller pod crash

[helm-template-crash-0.0.1.tar.gz](https://github.com/kubernetes/helm/files/1790435/helm-template-crash-0.0.1.tar.gz) has a minimal Helm chart that causes Tiller to panic; `recover` in the template rendering code seems like it would in fact fix the problem.

The chart contains a single configmap:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-{{ .Chart.Name }}-placeholder
data:
  release: {{ .Release.Name }}
  releaseButNotFoo: {{ list .Release.Name | without "foo" | join "" }}
  releaseButNotFooCorrect: {{ without (list .Release.Name) "foo" | join "" }}
```

The key error here is that `releaseButNotFoo` calls the spring `without` function with its parameters in the wrong order.  If you unpack the chart and run `helm template .` it panics and you get a Go backtrace, eventually pointing back at `k8s.io/helm/pkg/engine.(*Engine).render` (`k8s.io/helm/pkg/engine/engine.go:255`).  An earlier comment suggested `helm lint`, but that panics too.  If you actually try to `helm install` it, the tiller pod crashes, and `kubectl logs` says:

```
[tiller] 2018/03/07 18:39:37 rendering helm-template-crash chart using values
panic: Cannot find without on type string [recovered]
	panic: Cannot find without on type string

goroutine 245 [running]:
text/template.errRecover(0xc42082ae10)
	/usr/local/go/src/text/template/exec.go:146 +0x62
panic(0x18ba920, 0xc4204055b0)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
k8s.io/helm/vendor/github.com/Masterminds/sprig.without(0x18ba920, 0xc420405590, 0xc4204055a0, 0x1, 0x1, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/vendor/github.com/Masterminds/sprig/list.go:237 +0x3cb
reflect.Value.call(0x195f4a0, 0x1cf4a10, 0x13, 0x1c499d7, 0x4, 0xc420d71710, 0x2, 0x2, 0xc420d73080, 0x1, ...)
	/usr/local/go/src/reflect/value.go:434 +0x91f
reflect.Value.Call(0x195f4a0, 0x1cf4a10, 0x13, 0xc420d71710, 0x2, 0x2, 0x186f140, 0xc420d716f8, 0x97)
	/usr/local/go/src/reflect/value.go:302 +0xa4
text/template.(*state).evalCall(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x195f4a0, 0x1cf4a10, 0x13, 0x297f3e0, 0xc420d71350, 0xc420ddc774, ...)
	/usr/local/go/src/text/template/exec.go:668 +0x51f
text/template.(*state).evalFunction(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420d71380, 0x297f3e0, 0xc420d71350, 0xc420d72f20, 0x2, 0x2, ...)
	/usr/local/go/src/text/template/exec.go:536 +0x18e
text/template.(*state).evalCommand(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420d71350, 0x186f140, 0xc420d716f8, 0x97, 0x186f140, 0xc420d716f8, ...)
	/usr/local/go/src/text/template/exec.go:433 +0x6cd
text/template.(*state).evalPipeline(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0xc420a9d630, 0x0, 0x0, 0x98)
	/usr/local/go/src/text/template/exec.go:406 +0xf2
text/template.(*state).walk(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x297f260, 0xc420d71410)
	/usr/local/go/src/text/template/exec.go:232 +0x55a
text/template.(*state).walk(0xc42082ad90, 0x1af1f40, 0xc420d70bd0, 0x15, 0x297f5c0, 0xc420d70f60)
	/usr/local/go/src/text/template/exec.go:240 +0x139
text/template.(*Template).execute(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0x1af1f40, 0xc420d70bd0, 0x0, 0x0)
	/usr/local/go/src/text/template/exec.go:195 +0x20a
text/template.(*Template).Execute(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0x1af1f40, 0xc420d70bd0, 0xc420d71601, 0xc420405420)
	/usr/local/go/src/text/template/exec.go:178 +0x53
text/template.(*Template).ExecuteTemplate(0xc420d3f1c0, 0x295a0e0, 0xc4203a9030, 0xc420d70e40, 0x25, 0x1af1f40, 0xc420d70bd0, 0x1, 0x11d)
	/usr/local/go/src/text/template/exec.go:165 +0x77
k8s.io/helm/pkg/engine.(*Engine).render(0xc42031e320, 0xc420d70de0, 0xc420d70de0, 0x295a9e0, 0xc42031e320)
	/go/src/k8s.io/helm/pkg/engine/engine.go:255 +0xcc6
k8s.io/helm/pkg/engine.(*Engine).Render(0xc42031e320, 0xc420d5ec00, 0xc420d70bd0, 0xc42031e320, 0x1, 0xc42035c000)
	/go/src/k8s.io/helm/pkg/engine/engine.go:123 +0x6a
k8s.io/helm/pkg/tiller.(*ReleaseServer).renderResources(0xc4202dd230, 0xc420d5ec00, 0xc420d70bd0, 0xc420d70b10, 0xc420404e40, 0xc42038f9d3, 0x5, 0x100, 0x1, 0xc420d72d40, ...)
	/go/src/k8s.io/helm/pkg/tiller/release_server.go:266 +0x1aa
k8s.io/helm/pkg/tiller.(*ReleaseServer).prepareRelease(0xc4202dd230, 0xc420d5eba0, 0xc42038fa50, 0x1, 0x1)
	/go/src/k8s.io/helm/pkg/tiller/release_install.go:87 +0x21d
k8s.io/helm/pkg/tiller.(*ReleaseServer).InstallRelease(0xc4202dd230, 0x7f264d855630, 0xc420d0bf50, 0xc420d5eba0, 0xc4202dd230, 0x1c9d07f, 0xe)
	/go/src/k8s.io/helm/pkg/tiller/release_install.go:36 +0xff
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_InstallRelease_Handler.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0x33, 0xc420a9ce10, 0x83b701, 0x2a1fd50)
	/go/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:1228 +0x86
k8s.io/helm/vendor/github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0xc420d5dfa0, 0xc420d5dfc0, 0xc4200ee7f0, 0xc420d5df20, 0xc42077f960, 0x40fe28)
	/go/src/k8s.io/helm/vendor/github.com/grpc-ecosystem/go-grpc-prometheus/server_metrics.go:86 +0xdb
k8s.io/helm/pkg/tiller.newUnaryInterceptor.func1(0x7f264d855630, 0xc420d0bf50, 0x1b8ad00, 0xc420d5eba0, 0xc420d5dfa0, 0xc420d5dfc0, 0x0, 0xc42077f9d8, 0x40fe28, 0x50)
	/go/src/k8s.io/helm/pkg/tiller/server.go:60 +0xb5
k8s.io/helm/pkg/proto/hapi/services._ReleaseService_InstallRelease_Handler(0x1c03f40, 0xc4202dd230, 0x7f264d855630, 0xc420d0bf50, 0xc420a9cc80, 0x1cf4608, 0x0, 0x0, 0x2, 0x2)
	/go/src/k8s.io/helm/pkg/proto/hapi/services/tiller.pb.go:1230 +0x177
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500, 0xc4202dd320, 0x29eb8c8, 0x0, 0x0, 0x0)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:826 +0xc41
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).handleStream(0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500, 0x0)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:1023 +0x15a6
k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc4202acb80, 0xc4200ee780, 0x2981120, 0xc42083e160, 0xc420cf3500)
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:572 +0xa9
created by k8s.io/helm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/k8s.io/helm/vendor/google.golang.org/grpc/server.go:573 +0xa1
```

```% helm version
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
```

--- a/pkg/engine/engine.go
+++ b/pkg/engine/engine.go
@@ -19,6 +19,7 @@ package engine
 import (
 	"bytes"
 	"fmt"
+	"log"
 	"path"
 	"sort"
 	"strings"
@@ -204,6 +205,11 @@ func (e *Engine) render(tpls map[string]renderable) (map[string]string, error) {
 	// The idea with this process is to make it possible for more complex templates
 	// to share common blocks, but to make the entire thing feel like a file-based
 	// template engine.
+	defer func() {
+		if err := recover(); err != nil {
+			log.Printf("rendering template failed: %v\n", err)
+		}
+	}()
 	t := template.New("gotpl")
 	if e.Strict {
 		t.Option("missingkey=error")

--- a/pkg/engine/engine.go
+++ b/pkg/engine/engine.go
@@ -19,7 +19,6 @@ package engine
 import (
 	"bytes"
 	"fmt"
-	"log"
 	"path"
 	"sort"
 	"strings"
@@ -197,7 +196,7 @@ func (e *Engine) alterFuncMap(t *template.Template) template.FuncMap {
 }
 
 // render takes a map of templates/values and renders them.
-func (e *Engine) render(tpls map[string]renderable) (map[string]string, error) {
+func (e *Engine) render(tpls map[string]renderable) (rendered map[string]string, err error) {
 	// Basically, what we do here is start with an empty parent template and then
 	// build up a list of templates -- one for each file. Once all of the templates
 	// have been parsed, we loop through again and execute every template.
@@ -206,8 +205,8 @@ func (e *Engine) render(tpls map[string]renderable) (map[string]string, error) {
 	// to share common blocks, but to make the entire thing feel like a file-based
 	// template engine.
 	defer func() {
-		if err := recover(); err != nil {
-			log.Printf("rendering template failed: %v\n", err)
+		if r := recover(); r != nil {
+			err = fmt.Errorf("rendering template failed: %v", r)
 		}
 	}()
 	t := template.New("gotpl")
@@ -247,7 +246,7 @@ func (e *Engine) render(tpls map[string]renderable) (map[string]string, error) {
 		}
 	}
 
-	rendered := make(map[string]string, len(files))
+	rendered = make(map[string]string, len(files))
 	var buf bytes.Buffer
 	for _, file := range files {
 		// Don't render partials. We don't care out the direct output of partials.

